pkg/reader/usdc_reader.go

```diff
@@ -0,0 +1,282 @@
+package reader
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+
+	sel "github.com/smartcontractkit/chain-selectors"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/query"
+
+	"github.com/smartcontractkit/chainlink-ccip/execute/exectypes"
+	typconv "github.com/smartcontractkit/chainlink-ccip/internal/libs/typeconv"
+	"github.com/smartcontractkit/chainlink-ccip/pkg/consts"
+	"github.com/smartcontractkit/chainlink-ccip/pkg/contractreader"
+	"github.com/smartcontractkit/chainlink-ccip/pluginconfig"
+)
+
+type MessageHash []byte
+
+type USDCMessageReader interface {
+	MessageHashes(ctx context.Context,
+		source, dest cciptypes.ChainSelector,
+		tokens map[exectypes.MessageTokenID]cciptypes.RampTokenAmount,
+	) (map[exectypes.MessageTokenID]MessageHash, error)
+}
+
+const (
+	CCTPMessageVersion = uint32(0)
+)
+
+// TODO, this should be fetched from USDC Token Pool and cached
+var CCTPDestDomains = map[uint64]uint32{
+	sel.ETHEREUM_MAINNET.Selector:                    0,
+	sel.AVALANCHE_MAINNET.Selector:                   1,
+	sel.ETHEREUM_MAINNET_OPTIMISM_1.Selector:         2,
+	sel.ETHEREUM_MAINNET_ARBITRUM_1.Selector:         3,
+	sel.ETHEREUM_MAINNET_BASE_1.Selector:             6,
+	sel.POLYGON_MAINNET.Selector:                     7,
+	sel.ETHEREUM_TESTNET_SEPOLIA.Selector:            0,
```

Comment:
Does the token pool have same values for testnets as mainnets? can't this introduce some mix between mainnet and testnet?

Reply:
Token Pool doesn't, but CCTP does. TokenPool knows only about either mainnet or testnet. However, CCTP uses the same values for the particular chain, regarding on whether it's mainnet and testnet

Reply:
Meaning that Arbitrum Mainnet and Arbitrum Testnet has always domain set to 3

Reply:
Btw, these values are replicated from CCTP contracts, we only keep the mapping from our selectors to their domains, so there is not much we can do about conflicting values. This is how they implemented that 🤷